### PR TITLE
Jjjoness/asset browser table view#12862

### PR DESCRIPTION
**Pull request by jjjoness**
## What does this PR do?

This adds the new extended table view to the Asset Browser.

## How was this PR tested?
Tested locally.

image

### CODE REVIEW

**Review by dtamkin1 — APPROVED** [review]

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedTableViewProxyModel.cpp [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+            case ".font"_hash:
+                return "Font";
+            case ".xml"_hash:
+                return "XML";
+            case ".json"_hash:
+                return "JSON";
+            case ".exr"_hash:
+                return "EXR";
+            case ".wav"_hash:
+                return ".WAV";
```
Comment (cgalvan) [inline Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedTableViewProxyModel.cpp:180]: Should `WAV` have the `.` removed as well?

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedTableViewProxyModel.cpp [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+                return "Anim Graph";
+            case ".motionset"_hash:
+                return "Motion Set";
+            case ".assetinfo"_hash:
+                return "Asset Info";
+            case ".py"_hash:
+                return "Python Script";
+            case ".lua"_hash:
+                return "Lua Script";
+            case ".tif"_hash:
```
Comment (cgalvan) [inline Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedTableViewProxyModel.cpp:163]: Should handle `.tiff` as well

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/Entries/RootAssetBrowserEntry.cpp [hunk 3]
```diff
@@ -127,6 +130,33 @@ namespace AzToolsFramework
...
+                    FILE* f1;
+                    azfopen(&f1, assetPath.c_str(), "r");
+                    if (f1)
+                    {
+                        char readBuffer[65536];
+                        rapidjson::FileReadStream is(f1, readBuffer, sizeof(readBuffer));
+
+                        rapidjson::Document doc;
+                        doc.ParseStream(is);
+                        fclose(f1);
```
Comment (cgalvan) [inline Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/Entries/RootAssetBrowserEntry.cpp:148]: Replace with one of the JSON reading helpers, such as `JsonSerializationUtils::ReadJsonFile`

File: Gems/SceneProcessing/Code/Source/SceneBuilder/SceneBuilderWorker.cpp [hunk 4]
```diff
@@ -435,6 +436,33 @@ namespace SceneBuilder
...
+            writer.Double(dimension.GetX());
+            writer.Double(dimension.GetY());
+            writer.Double(dimension.GetZ());
+            writer.EndArray();
+            writer.EndObject();
+            writer.EndObject();
+            FILE* f1;
+            azfopen(&f1, folder.c_str(), "w");
+            fwrite(s.GetString(), 1, s.GetSize(), f1);
+            fclose(f1);
```
Comment (cgalvan) [inline Gems/SceneProcessing/Code/Source/SceneBuilder/SceneBuilderWorker.cpp:462]: Replace with `JsonSerializationUtils::WriteJsonFile`

**Review by cgalvan — CHANGES_REQUESTED** [review]

File: Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/AssetFolderExpandedTableView.cpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+#include <QMouseEvent>
+#include <QPainter>
+#include <QScrollBar>
+#include <QSettings>
+#include <QMenu>
+AZ_POP_DISABLE_WARNING
+
+namespace AzQtComponents
+{
+    AssetFolderExpandedTableView::AssetFolderExpandedTableView(QWidget* parent)
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/AssetFolderExpandedTableView.cpp:23]: There was at one time AssetFolderTableView so created AssetFolderExpandedTableView. The old AssetFolderTableView was removed and this code went unchanged.

File: Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/AssetFolderExpandedTableView.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+namespace AzQtComponents
+{
+    class AZ_QT_COMPONENTS_API AssetFolderExpandedTableView
+        : public TableView
+    {
+        Q_OBJECT
+
+    public:
+        explicit AssetFolderExpandedTableView(QWidget* parent = nullptr);
+        ~AssetFolderExpandedTableView() override;
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/AssetFolderExpandedTableView.h:28]: Done.

File: Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/AssetFolderExpandedTableView.cpp [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        : TableView(parent)
+    {
+        setSortingEnabled(true);
+    }
+
+    AssetFolderExpandedTableView::~AssetFolderExpandedTableView() = default;
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/AssetFolderExpandedTableView.cpp:29]: Done.

File: Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/AssetFolderExpandedTableView.h [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+        void SetShowSearchResultsMode(bool searchMode);
+        void setRootIndex(const QModelIndex& index) override;
+
+    signals:
+        void rootIndexChanged(const QModelIndex& idx);
+        void showInFolderTriggered(const QModelIndex& idx);
+
+    private:
+        bool m_showSearchResultsMode = false;
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/AssetFolderExpandedTableView.h:38]: There is more work to be done in here to bring all the UX features in over the next few weeks, so I would prefer to keep this.

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedFilterModel.cpp [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        //////////////////////////////////////////////////////////////////////////
+        //AssetBrowserFilterModel
+        AssetBrowserExpandedFilterModel::AssetBrowserExpandedFilterModel(QObject* parent)
+            : AssetBrowserFilterModel(parent)
+        {
+            m_shownColumns.insert(aznumeric_cast<int>(AssetBrowserEntry::Column::Type));
+            m_shownColumns.insert(aznumeric_cast<int>(AssetBrowserEntry::Column::DiskSize));
+            m_shownColumns.insert(aznumeric_cast<int>(AssetBrowserEntry::Column::Vertices));
+            m_shownColumns.insert(aznumeric_cast<int>(AssetBrowserEntry::Column::ApproxSize));
+            m_shownColumns.insert(aznumeric_cast<int>(AssetBrowserEntry::Column::SourceControlStatus));
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedFilterModel.cpp:38]: Done.

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedFilterModel.h [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+        class AssetBrowserExpandedFilterModel
+            : public AssetBrowserFilterModel
+         {
+            Q_OBJECT
+
+        public:
+            AZ_CLASS_ALLOCATOR(AssetBrowserExpandedFilterModel, AZ::SystemAllocator, 0);
+            explicit AssetBrowserExpandedFilterModel(QObject* parent = nullptr);
+            ~AssetBrowserExpandedFilterModel() override;
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedFilterModel.h:28]: Done.

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedFilterModel.cpp [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+        AssetBrowserExpandedFilterModel::~AssetBrowserExpandedFilterModel()
+        {
+        }
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedFilterModel.cpp:43]: Done.

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/Views/AssetBrowserExpandedTableView.h [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+            void paint(QPainter* painter, const QStyleOptionViewItem& option, const QModelIndex& index) const override;
+        };
+
+        class AssetBrowserExpandedTableView
+            : public QWidget
+        {
+            Q_OBJECT
+        public:
+            AZ_CLASS_ALLOCATOR(AssetBrowserExpandedTableView, AZ::SystemAllocator, 0);
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/Views/AssetBrowserExpandedTableView.h:48]: Done.

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedFilterModel.h [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        using ShownColumnsSet = AZStd::fixed_unordered_set<int, 3, aznumeric_cast<int>(AssetBrowserEntry::Column::Count)>;
+
+        class AssetBrowserExpandedFilterModel
+            : public AssetBrowserFilterModel
+         {
+            Q_OBJECT
+
+        public:
+            AZ_CLASS_ALLOCATOR(AssetBrowserExpandedFilterModel, AZ::SystemAllocator, 0);
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedFilterModel.h:26]: Done.

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedTableViewProxyModel.cpp [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+
+        AssetBrowserExpandedTableViewProxyModel::~AssetBrowserExpandedTableViewProxyModel() = default;
+
+        QVariant AssetBrowserExpandedTableViewProxyModel::data(const QModelIndex& index, int role) const
+        {
+            auto assetBrowserEntry = mapToSource(index).data(AssetBrowserModel::Roles::EntryRole).value<const AssetBrowserEntry*>();
+            AZ_Assert(assetBrowserEntry, "Couldn't fetch asset entry for the given index.");
+            if (!assetBrowserEntry)
+            {
+                return " No Data ";
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedTableViewProxyModel.cpp:31]: Done.

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedTableViewProxyModel.cpp [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+                            switch (assetBrowserEntry->GetEntryType())
+                            {
+                            case AssetBrowserEntry::AssetEntryType::Root:
+                                return "Root";
+                            case AssetBrowserEntry::AssetEntryType::Folder:
+                                return "Folder";
+                            case AssetBrowserEntry::AssetEntryType::Source:
+                                return ExtensionToType(static_cast<const SourceAssetBrowserEntry*>(assetBrowserEntry)->GetExtension()).c_str();
+                            case AssetBrowserEntry::AssetEntryType::Product:
+                                return "Product";
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedTableViewProxyModel.cpp:53]: Done.

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedTableViewProxyModel.cpp [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+                            if (abs(dim.GetX())<1.0f && abs(dim.GetY())<1.0f && abs(dim.GetZ())<1.0f)
+                                return QString{ "%1 x %2 x %3" }.arg(dim.GetX()).arg(dim.GetY()).arg(dim.GetZ());
+                            return QString{ "%1 x %2 x %3" }
+                                .arg(static_cast<int>(dim.GetX()))
+                                .arg(static_cast<int>(dim.GetY()))
+                                .arg(static_cast<int>(dim.GetZ()));
+                        }
+                        return "";
+                    default:
+                        return "No data";
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedTableViewProxyModel.cpp:75]: Done.

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedTableViewProxyModel.cpp [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+            if (parent != m_rootIndex)
+            {
+                return false;
+            }
+            return true;
+        }
+
+        int AssetBrowserExpandedTableViewProxyModel::columnCount([[maybe_unused]]const QModelIndex& parent) const
+        {
+            return 6;
```
[the remark this answers is not in the document]
Comment (jjjoness) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedTableViewProxyModel.cpp:114]: Done.

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedTableViewProxyModel.cpp (hunk 1 above)
Comment (jjjoness) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedTableViewProxyModel.cpp:180]: Yes. Done.

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedTableViewProxyModel.cpp (hunk 2 above)
Comment (jjjoness) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserExpandedTableViewProxyModel.cpp:163]: Done.

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/Entries/RootAssetBrowserEntry.cpp (hunk 3 above)
Comment (jjjoness) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/Entries/RootAssetBrowserEntry.cpp:148]: Done.

File: Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/AssetFolderExpandedTableView.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (cgalvan) [reply Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/AssetFolderExpandedTableView.cpp]: Makes sense, thanks!

**Review by cgalvan — APPROVED** [review]

**Review by dtamkin1 — APPROVED** [review]

**Comment by rainbj** [comment]
I'm excited about a table view :)

**Pull request merged**